libbpf-tools/execsnoop.bpf.c

```diff
@@ -0,0 +1,142 @@
+
+// null isn't defined
+#define NULL 0
+// from linux/bpf.h
+#define BPF_NOEXIST  1
+
+#define BASE_EVENT_SIZE (sizeof(struct event) - FULL_MAX_ARGS_ARR)
+#define EVENT_SIZE(X) (BASE_EVENT_SIZE + (ARGSIZE * (X)->args_count))
+
+const volatile bool ignore_failed = true;
+const volatile uid_t targ_uid = INVALID_UID;
+const volatile int max_args = DEFAULT_MAXARGS;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct event);
+	__uint(max_entries, 1);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, pid_t);
+	__type(value, struct event);
+} stored SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline bool valid_uid(uid_t uid) {
+	return uid != INVALID_UID;
+}
+
+SEC("tracepoint/syscalls/sys_enter_execve")
+int tracepoint__syscalls__sys_enter_execve(struct trace_event_raw_sys_enter* ctx)
+{
+	pid_t pid;
+	struct event *event;
+	int ret;
+
+	uid_t uid = bpf_get_current_uid_gid() & 0xffffffff;
+
+	if (valid_uid(targ_uid) && targ_uid != uid)
+		return 0;
+
+	ret = 0;
+	event = bpf_map_lookup_elem(&start, &ret);
+
+	if (!event)
+		return 0;
+
+	pid = bpf_get_current_pid_tgid() >> 32;
+	event->pid = pid;
+	event->uid = uid;
+	event->more_args = false;
+	event->args_count = 0;
+
+	ret = bpf_probe_read_str(event->args, ARGSIZE, (const char*)ctx->args[0]);
+	if (ret < 0)
+		goto out;
+
+	event->args_count++;
+	const char **args = (const char **)(ctx->args[1]);
+
+	#pragma unroll
+	for (int i = 1; i < TOTAL_MAX_ARGS && i < max_args; i++) {
+		const char *argp = NULL;
+		char *argv_write = &(event->args[ARGSIZE * i]);
+
+		bpf_probe_read(&argp, sizeof(argp), &args[i]);
+
+		if (!argp)
+			goto out;
+
+		ret = bpf_probe_read_str(argv_write, ARGSIZE, (void *)argp);
+		if (ret < 0)
+			goto out;
+
+		event->args_count++;
+	}
+
+	event->more_args = true;
+
+out:
+	bpf_map_update_elem(&stored, &pid, event, BPF_NOEXIST);
+	return 0;
+}
+
+SEC("tracepoint/syscalls/sys_exit_execve")
+int tracepoint__syscalls__sys_exit_execve(struct trace_event_raw_sys_exit* ctx)
+{
+	pid_t pid;
+	int ret;
+	struct event *event;
+	struct task_struct *task;
+	struct task_struct *parent;
+
+	pid = bpf_get_current_pid_tgid() >> 32;
+
+	ret = ctx->ret;
+	if (ignore_failed && ret < 0)
+		goto out;
+
+	u32 uid = (u32)bpf_get_current_uid_gid();
+	if (valid_uid(targ_uid) && targ_uid != uid)
+		goto out;
+
+	event = bpf_map_lookup_elem(&stored, &pid);
+
+	if (!event)
+		goto out;
+
+	event->pid = pid;
+	event->uid = uid;
+	event->retval = ret;
+
+	task = (struct task_struct*)bpf_get_current_task();
+	bpf_probe_read(&parent, sizeof(parent), &task->real_parent);
+	bpf_probe_read(&event->ppid, sizeof(pid_t), &parent->tgid);
+	bpf_get_current_comm(&event->comm, sizeof(event->comm));
+
+	/* trick to satisfy verifier */
+	#pragma unroll
```

Comment:
Had to do this trick to satisfy verifier, after loop unrolling size argument (last argument) should be constant.